#98441 - Ignore invalid handle/access when setting console cp

nick-beer opened the pull request:
There are situations where setting the console code page is expected to fail. For instance, if the application was started with DETACHED_PROCESS, the attempting to set the code page will fail with ERROR_INVALID_HANDLE. In these cases, ignore the returned error.

Fixes #98441

nick-beer committed:
Ignore invalid handle/access when setting console cp

There are situations where setting the console code page is expected to
fail. For instance, if the application was started with DETACHED_PROCESS,
the attempting to set the code page will fail with ERROR_INVALID_HANDLE.
In these cases, ignore the returned error.

Fix #98441

nick-beer commented:
@dotnet-policy-service agree company="National Instruments"

jkotas commented on src/libraries/System.Console/src/System/ConsolePal.Windows.cs:
```diff
@@ -122,10 +124,25 @@ public static void SetConsoleOutputEncoding(Encoding enc)
 
+        private static void HandleSetConsoleEncodingError()
+        {
+            var lastError = Marshal.GetLastPInvokeError();
```
```suggestion
            int lastError = Marshal.GetLastPInvokeError();
```
The [repo coding conventions](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) do not allow `var` to be used in this position.

jkotas commented:
Is it possible to add a test?

nick-beer committed:
Update src/libraries/System.Console/src/System/ConsolePal.Windows.cs

coding conventions: use explicit Type rather than var

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

danmoseley commented on src/libraries/System.Console/src/System/ConsolePal.Windows.cs:
```diff
@@ -122,10 +124,25 @@ public static void SetConsoleOutputEncoding(Encoding enc)
 
+        private static void HandleSetConsoleEncodingError()
+        {
+            int lastError = Marshal.GetLastPInvokeError();
```
It may be better to pass the last error into this method. It would be easy for future refactoring to adda line of code before the call to this method that resets last error. Better to store it immediately after it is set.

nick-beer committed:
Pass last error into error handling function

This helps ensure GetLastError is called immediately after the failing
win32 function.

jhudsoncedaron commented:
@jkotas : Yes, but only if the test is allowed to start a new project. I put a short crash sequence in the work item thread.

nick-beer commented:
Yeah - I should have updated this thread.  I'm currently working on a couple tests for this issue. It took a little bit of time to aquaint myself with the test infrastrucure.  It looks to me like the existing `RemoteExecutor` infrastructure only uses `Process.Start` to run the external process, so it's not easy to start a detached process.  I now have tests that use `FreeConsole` to simulate a similar scenario, but I'm getting unexpected results.  Trying to debug the failures is resulting in Visual Studio crashes, so it's taking a little longer than I'd hoped.

Do we agree that tests with `FreeConsole` to simulate a detached console are valuable?

jkotas commented:
Either `FreeConsole` or launching a process with `DETACHED_PROCESS` flags should work for a test.

If you would like to go with `DETACHED_PROCESS`, [ProcessorCount_Windows_RespectsJobCpuRateAndConfigurationSetting](https://github.com/dotnet/runtime/blob/5568ac9e1a7215e9ac1c84f0c79c7da8b16d1e50/src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.ProcessorCount.cs#L117-L133) has an example for how to launch child test process with custom Windows-specific settings.

nick-beer committed:
Add tests validating new behavior

nick-beer commented:
Tests added.  I went with FreeConsole.  I like the way the test you linked to re ran the process with custom flags - pretty cool.

jhudsoncedaron commented on src/libraries/System.Console/tests/ConsoleEncoding.Windows.cs:
```diff
@@ -38,6 +38,27 @@ public void InputEncoding_SetUnicodeEncoding_SilentlyIgnoredInternally()
+            // The internal state of Console should have updated, despite the failure to change the console's input encoding
+            Assert.Equal(encoding, Console.InputEncoding);
+            // Operations on the console are no longer valid - GetConsoleCP fails.
+            Assert.Equal(0u, GetConsoleCP());
```
I would advise testing that Console.WriteLine() still works here. The caller would need to use StandardOutputRedirected = true for the test case.

jkotas reviewed (approved):
LGTM! Thank you

jkotas commented:
The test is failing on Windows Nano Server:
```
Unhandled exception. Xunit.Sdk.NotEqualException: Assert.NotEqual() Failure: Values are equal
Expected: Not 65001
Actual:       65001
   at Xunit.Assert.NotEqual[T](T expected, T actual, IEqualityComparer`1 comparer) in /_/src/Microsoft.DotNet.XUnitAssert/src/EqualityAsserts.cs:line 696
   at Xunit.Assert.NotEqual[T](T expected, T actual) in /_/src/Microsoft.DotNet.XUnitAssert/src/EqualityAsserts.cs:line 610
   at ConsoleEncoding.<>c.<InputEncoding_SetEncodingWhenDetached_ErrorIsSilentlyIgnored>b__12_0() in /_/src/libraries/System.Console/tests/ConsoleEncoding.Windows.cs:line 48
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args) in /_/src/coreclr/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.CoreCLR.cs:line 36
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs:line 57
--- End of stack trace from previous location ---
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 97
    ConsoleEncoding.InputEncoding_SetEncodingWhenDetached_ErrorIsSilentlyIgnored [FAIL]
      Microsoft.DotNet.RemoteExecutor.RemoteExecutionException : Remote process failed with an unhandled exception.
      Stack Trace:
        
        Child exception:
          Xunit.Sdk.NotEqualException: Assert.NotEqual() Failure: Values are equal
        Expected: Not 65001
        Actual:       65001
        /_/src/libraries/System.Console/tests/ConsoleEncoding.Windows.cs(48,0): at ConsoleEncoding.<>c.<InputEncoding_SetEncodingWhenDetached_ErrorIsSilentlyIgnored>b__12_0()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
        /_/src/coreclr/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.CoreCLR.cs(36,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(57,0): at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
        
        Child process:
          System.Console.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 ConsoleEncoding+<>c Void <InputEncoding_SetEncodingWhenDetached_ErrorIsSilentlyIgnored>b__12_0()
```

jhudsoncedaron commented:
I see why the test is failing. The code asserts that the current output codepage is not the one it's going to try to test by changing it to. So the test knows it doesn't work and bails.

Looks like the test harness needs to handle that case. Clearly the implementation needs to be if already 65001 use some other codepage instead of 65001. I'm not sure which code page it should change to though. Code page 437 would be a sane choice everywhere else but I'm not sure about nano.

jkotas commented:
Any one of the built-in .NET encodings should work: https://github.com/dotnet/runtime/blob/1b1d26ac74e374309fb101382026ba0480d3821d/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs#L208-L215

nick-beer committed:
Ensure we set an encoding that is different than the existing encoding

nick-beer commented:
Updated to ensure we set an encoding that's different from whatever is existing.  Also updated two existing tests that were not performing such a check.

danmoseley commented on src/libraries/System.Console/tests/ConsoleEncoding.Windows.cs:
```diff
@@ -38,13 +41,38 @@ public void InputEncoding_SetUnicodeEncoding_SilentlyIgnoredInternally()
         RemoteExecutor.Invoke(() =>
         {
-            Encoding encoding = Encoding.GetEncoding(0);
+            Encoding encoding = Console.OutputEncoding.CodePage != Encoding.ASCII.CodePage
```
This test seems to be passing 0 specifically to get the system's default encoding, even if it's the current encoding

nick-beer committed:
Revert 'default encoding' tests to use default even if that's the current encoding

adamsitnik reviewed (approved):
LGTM, thank you @nick-beer for providing not just the fix but also great tests!
